### DISCUSSION TITLE
[Blitz Internals] Caching and not-drilling down server config

### DISCUSSION THREAD

**Discussion by svobik7** [blitz-js/blitz#941] [Ideas] [answered]
As me and @ryardley came to the idea of not-drilling down the server config in there: https://github.com/blitz-js/blitz/pull/878#discussion_r476264197

I am opening this threat to discuss if it make sense to create some global place from where the `ServerConfig` can be easily obtained and avoid passing config params down on multiple places.

My idea is something like this (not tested just came to my mind now):

- Config params should be calculated just once and then cached
- Config params can be forced to recalculate
- There is no need to await for both next-bins

```ts
let cachedConfig: NormalizedConfig | null = null

export async function getNormalizedConfig(
  config: ServerConfig,
  useCache: boolean = true,
): Promise<NormalizedConfig> {
  
  if (!useCache || cachedConfig === null) {
   
    // maybe use `config.rootFolder || pkgDir.sync() || process.cwd()`, is it there any benefit using pkgDir?
    const rootFolder = config.rootFolder || process.cwd();

    // do not await for both bin-pkg because always just one is used
    const nextBinPkg = config.interceptNextErrors ? "@blitzjs/server" : "next"
    const nextBinExec = config.interceptNextErrors ? "next-patched" : undefined
    const nextBin = await resolveBinAsync(nextBinPkg, nextBinExec)

    const git = parseChokidarRulesFromGitignore(resolve(process.cwd(), rootFolder))
    const isTypescript = fs.existsSync(path.join(rootFolder, "tsconfig.json"))

    cachedConfig = {
      ...config,
      buildFolder: resolve(rootFolder, config.buildFolder ?? defaults.buildFolder),
      devFolder: resolve(rootFolder, config.devFolder ?? defaults.devFolder),
      ignore: defaults.ignoredPaths.concat(git.ignoredPaths),
      include: defaults.includePaths.concat(git.includePaths),
      nextBin: resolve(rootFolder, nextBin),
      transformFiles: config.transformFiles ?? transformFiles,
      watch: config.watch ?? false,
      writeManifestFile: config.writeManifestFile ?? defaults.writeManifestFile,
      isTypescript: isTypescript ?? true,
    }
  }

  return cachedConfig;
}
```

I think it would be easily to get current server config with `const config = getNormalizedConfig()` anywhere in server package.

How does it sound to others?

**Comment by flybayer** [comment]
Sounds like a good idea to me!

**Comment by svobik7** [comment]
Cool. Is that something I should start the working on or should I wait for more opinions? :-)

**Reply by flybayer** [reply]
Go for it :)

**Comment by svobik7** [comment, marked as answer]
After a small time of investigating in this I came up to resolution that initial idea of caching the server config leads to no benefits. It breaks the "pure" concept of `normalize` function in server package and makes future testing hard to maintain. So therefore I do not think that server config caching is benefical here any more.

Instead of caching I created PR with small server code refactor and made possible to skip TS checking by explicitly provide new flag `--ts` --> https://github.com/blitz-js/blitz/pull/1014

**Reply by flybayer** [reply]
Great, thank you for your investigation and work on this!
